### ISSUE TITLE
Allow for ASL modification without a full infra update.

### ISSUE THREAD

**Issue by ljacobsson** [aws/aws-sam-cli#4408]
Currently an update to an ASL file triggers a full infra update. 

I think the intended behaviour is to make an on-the-air update to swap out the SateMachine's ASL.

![ezgif-5-11d65ab977](https://user-images.githubusercontent.com/7579097/201709068-340b3f0a-ac30-4d5a-bd2f-3c766527601d.gif)

**Comment by qingchm** [comment]
Hey @ljacobsson actually normal state machine definition updates will go through code sync instead of the infra sync, but unfortunately it looks to me that your template involve DefinitionSubstitutions if I'm not guessing wrong? If a DefinitionSubstitutions is provided, we currently will have to go through infra sync. We will recognize this pain point and consider ways to support DefinitionSubstitutions related code syncs! Feel free to use this issue as a contact point to poke me for any updates!

**Comment by ljacobsson** [comment]
That's right that it's using DefinitionSubstitutions and after a chat with @praneetap I understand that this is the current behaviour.

This is a feature request and not a bug report though. It makes a lot of sense to allow for quick iterations even if substitutions are used. I can't think of any state machine I've built that doesn't use substitutions.

I've solved it in one of my side projects, [sfn-cli](https://github.com/mhlabs/sfn-cli#real-time-sync-your-local-asl-with-the-cloud).

Also, see [this thread](https://twitter.com/lajacobsson/status/1593628492146548738)

**Comment by qingchm** [comment]
Great that you have some work done to help with supporting DefinitionSubstitutions, would you be interested to investigate on making a contribution to our sync logic for this then? Meanwhile I will discuss internally on where this feature lies on the sync related improvements that our team plans to improve on!

As of right now the limitation is that each sync flow is a poller on file changes on their own, and for template changes, we will queue up an infra sync task. The solution would likely be that before we send the infra sync task, we parse the template and detect if it's DefinitionSubstitutions related change, if so we submit an stepfunction sync flow instead. The stepfunction sync flow should follow your logic here https://github.com/mhlabs/sfn-cli/blob/main/src/commands/sync/sync.js#L63 to use both files for an ASL update. Again appreciate that you reached out for us to know the need for DefinitionSubstitutions being supported, and community contribution is always more than welcomed if you would like to!

**Comment by qingchm** [comment]
@ljacobsson also assigning myself here for this sync related issue, feel free to contact me here for any updates!
